File: framework/languages/java/samples/java/GameQuest/Server/QuestMission/src/main/java/systems/zlink/samples/gamequest/server/questmission/handlers/GameplayEventHandler.java
package systems.zlink.samples.gamequest.server.questmission.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.gamequest.server.questmission.store.QuestStore;
import systems.zlink.samples.gamequest.shared.contracts.Messages;

@ZLinkHandlerGroup("quest-owner")
public final class GameplayEventHandler
    implements ZLinkRequestHandler<Messages.GameplayEventEnvelope, Messages.QuestProcessingRes> {
    private final QuestStore store;

    public GameplayEventHandler(QuestStore store) {
        this.store = store;
    }

    @Override
    public Messages.QuestProcessingRes handle(
        Messages.GameplayEventEnvelope request,
        ZLinkRequestContext context) {
        store.markRehydrated(request.playerId());
        return store.apply(request);
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/QuestMission/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/questmission/handlers/GameplayEventHandler.kt
package systems.zlink.samples.kotlin.gamequest.server.questmission.handlers

import systems.zlink.framework.channels.ZLinkPublishContext
import systems.zlink.framework.kotlin.ZLinkSuspendingPublishHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.gamequest.server.configuration.SampleNames
import systems.zlink.samples.kotlin.gamequest.server.questmission.application.QuestEventProcessor
import systems.zlink.samples.kotlin.gamequest.shared.contracts.GameplayEventMsg

/** Fanout subscriber: processes published gameplay events. Mirrors the .NET `GameplayEventHandler`. */
@ZLinkHandlerGroup("gamequest-gameplay")
class GameplayEventHandler(
    private val processor: QuestEventProcessor,
) : ZLinkSuspendingPublishHandler<GameplayEventMsg> {
    override suspend fun handle(message: GameplayEventMsg, context: ZLinkPublishContext) = run {
        if (context.topic() != SampleNames.GameplayTopic) {
            return
        }
        processor.process(message)
    }
}
